File: javascript/colas/colaLista.js
const readline = require('readline');

class Node {
    constructor() {
        this.data = 0;
        this.next = null;
    }
}

let head = null;
let tail = null;

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function insertar() {
    rl.question('\nIngrese el elemento: ', (input) => {
        const elemento = parseInt(input);
        
        const ptr = new Node();
        
        if (ptr == null) {
            console.log('\nDESBORDAMIENTO (OVERFLOW)\n');
            mostrarMenu();
            return;
        }
        
        if (head == null && tail == null) {
            head = tail = ptr;
        } else {
            tail.next = ptr;
            tail = ptr;
        }
        
        ptr.data = elemento;
        console.log('\nElemento insertado correctamente.\n');
        mostrarMenu();
    });
}

function eliminar() {
    if (head == null) {
        console.log('\nSUBDESBORDAMIENTO (UNDERFLOW)\n');
        mostrarMenu();
        return;
    }
    
    const elemento = head.data;
    
    if (head == tail) {
        head = tail = null;
    } else {
        head = head.next;
    }
    
    console.log('\nElemento eliminado: ' + elemento + '\n');
    mostrarMenu();
}

function mostrar() {
    let ptr = head;
    
    if (ptr == null) {
        console.log('\nNada que imprimir');
    } else {
        process.stdout.write('\nImprimiendo valores . . . . . . \n');
        while (ptr != null) {
            process.stdout.write(ptr.data.toString());
            if (ptr != tail) {
                process.stdout.write(' | ');
            }
            ptr = ptr.next;
        }
        console.log();
    }
    mostrarMenu();
}

function mostrarMenu() {
    console.log('\n****************** MENÚ PRINCIPAL ******************\n');
    console.log('====================================================\n');
    console.log('1. Insertar un elemento\n');
    console.log('2. Eliminar un elemento\n');
    console.log('3. Mostrar la cola\n');
    console.log('4. Salir\n');
    
    rl.question('Ingrese su opción: ', (opcion) => {
        const opt = parseInt(opcion);
        
        switch (opt) {
            case 1:
                insertar();
                break;
            case 2:
                eliminar();
                break;
            case 3:
                mostrar();
                break;
            case 4:
                console.log('\nSaliendo del programa...\n');
                rl.close();
                break;
            default:
                console.log('\nOpción inválida. Intente nuevamente.\n');
                mostrarMenu();
        }
    });
}

// Iniciar el programa
mostrarMenu();

File: python/colas/colaLista.py
class Node:
    def __init__(self):
        self.data = 0
        self.next = None

head = None
tail = None

def insertar():
    global head, tail
    
    elemento = int(input("\nIngrese el elemento: "))
    
    ptr = Node()
    
    if ptr is None:
        print("\nDESBORDAMIENTO (OVERFLOW)\n")
        return
    
    if head is None and tail is None:
        head = tail = ptr
    else:
        tail.next = ptr
        tail = ptr
    
    ptr.data = elemento
    print("\nElemento insertado correctamente.\n")

def eliminar():
    global head, tail
    
    if head is None:
        print("\nSUBDESBORDAMIENTO (UNDERFLOW)\n")
        return
    
    elemento = head.data
    
    if head == tail:
        head = tail = None
    else:
        head = head.next
    
    print(f"\nElemento eliminado: {elemento}\n")

def mostrar():
    ptr = head
    
    if ptr is None:
        print("\nNada que imprimir")
    else:
        print("\nImprimiendo valores . . . . . . ")
        while ptr is not None:
            print(ptr.data, end='')
            if ptr != tail:
                print(' | ', end='')
            ptr = ptr.next
        print()

def main():
    opcion = 0
    
    while opcion != 4:
        print("\n****************** MENÚ PRINCIPAL ******************\n")
        print("====================================================\n")
        print("1. Insertar un elemento\n")
        print("2. Eliminar un elemento\n")
        print("3. Mostrar la cola\n")
        print("4. Salir\n")
        
        opcion = int(input("Ingrese su opción: "))
        
        if opcion == 1:
            insertar()
        elif opcion == 2:
            eliminar()
        elif opcion == 3:
            mostrar()
        elif opcion == 4:
            print("\nSaliendo del programa...\n")
        else:
            print("\nOpción inválida. Intente nuevamente.\n")

if __name__ == "__main__":
    main()